### PR TITLE
[5.3] Add findBy and findByOrFail methods

### PR DESCRIPTION
**Pull request by andrewdavis-dev**
I would like to add a findBy method that would work similarly to find, except it allows you to specify a column. For example, if you have a User model, you will often write the following code to get a single user:
```
$user = User::find(1);
```
But what if you need to find a single user by a username?
```
$user = User::where('username', 'jdoe')->first();
```
What I'm proposing is just a convenience method to make the above more concise:
```
$user = User::findBy('username', 'jdoe');
```

The method will just always return the first result found by searching the column with the specified value. I included findByOrFail to be consistent.

Thanks!

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -212,6 +227,26 @@ public function findOrFail($id, $columns = ['*'])
...
+     * @throws \Illuminate\Database\Eloquent\ModelNotFoundException
+     */
+    public function findByOrFail($column, $value, $columns = ['*'])
+    {
+        $result = $this->findBy($column, $value, $columns);
+        if(! is_null($result)) {
+            return $result;
+        }
+
+        throw (new ModelNotFoundException)->setModle(get_class($this->model), $id);
```
Comment (ntzm) [inline src/Illuminate/Database/Eloquent/Builder.php]: *setModel

Can you make some tests for these methods?

**Comment by restored18** [comment]
Okay, got the tests added and all the checks are passing. Let me know if you need anything else from me.

**Comment by JosephSilber** [comment]
I'm very much against this.

Doesn't really add anything, and `findByOrFail` is a nasty method name.

**Comment by taylorotwell** [comment]
Will hold off on this for now.

**Comment by restored18** [comment]
@JosephSilber If you go by that thinking then we should get rid of the find() method as well. I have seen a findBy method in other active record implementations. It makes simple finds more concise which is always a win.

@taylorotwell Can you explain what you don't like about it?

**Comment by fernandobandeira** [comment]
Maybe changing it to ``` findByField ``` and ``` findByFieldOrFail ``` would help a little with the method name.

**Comment by garygreen** [comment]
Can we also get a `findWhereNotOnATuesdayWhenEatenBeansOnToastForDinner` pls

**Pull request closed without merging**